compiler/lib/effects.ml

```diff
@@ -0,0 +1,416 @@
+            [ Let (k', Prim (Extern "caml_resume_stack", [ Pv stack; Pv k ]))
+            ; Let (x, Apply { f; args = [ arg; k' ]; exact = false })
+            ])
+    | Prim (Extern "%perform", [ Pv effect ]) ->
+        Some
+          (fun ~x ~k ->
+            [ Let
+                (x, Prim (Extern "caml_perform_effect", [ Pv effect; Pc (Int 0l); Pv k ]))
+            ])
+    | Prim (Extern "%reperform", [ Pv eff; Pv continuation ]) ->
+        Some
+          (fun ~x ~k ->
+            [ Let
+                (x, Prim (Extern "caml_perform_effect", [ Pv eff; Pv continuation; Pv k ]))
+            ])
+    | _ -> None
+  in
+
+  let rewritten_block =
+    match List.split_last block.body, block.branch with
+    | Some (body_prefix, Let (x, e)), Return ret ->
+        Option.map (rewrite_instr e) ~f:(fun instrs ->
+            assert (List.is_empty alloc_jump_closures);
+            assert (Var.equal x ret);
+            body_prefix, instrs ~x ~k, block.branch)
+    | Some (body_prefix, Let (x, e)), Branch cont ->
+        let allocate_continuation f =
+          let constr_cont, k' =
+            (* Construct continuation: it binds the return value [x],
+               allocates closures for dominated blocks and jumps to the
+               next block. *)
+            let pc, args = cont in
+            let ret = Var.fresh () in
+            let f' = closure_of_pc ~st pc in
+            allocate_closure
+              ~st
+              ~params:[ x ]
+              ~body:
+                (alloc_jump_closures @ [ Let (ret, Apply { f = f'; args; exact = true }) ])
+              ~branch:(Return ret)
+          in
+          let ret = Var.fresh () in
+          body_prefix, constr_cont @ f ~x:ret ~k:k', Return ret
+        in
+        Option.map (rewrite_instr e) ~f:allocate_continuation
+    | Some (_, (Set_field _ | Offset_ref _ | Array_set _ | Assign _)), _
+    | Some _, (Raise _ | Stop | Cond _ | Switch _ | Pushtrap _ | Poptrap _)
+    | None, _ -> None
+  in
+
+  let body, last =
+    match rewritten_block with
+    | Some (body_prefix, last_instrs, last) ->
+        List.map body_prefix ~f:(fun i -> cps_instr ~st i) @ last_instrs, last
+    | None ->
+        let last_instrs, last = cps_last ~st block.branch ~k in
+        let body =
+          List.map block.body ~f:(fun i -> cps_instr ~st i)
+          @ alloc_jump_closures
+          @ last_instrs
+        in
+        body, last
+  in
+
+  { params = block.params; body; branch = last }
+
+let split_blocks (p : Code.program) =
+  (* Ensure that function applications and effect primitives are in
+     tail position *)
+  let split_block pc block p =
+    let is_split_point i r branch =
+      match i with
+      | Let (x, (Apply _ | Prim (Extern ("%resume" | "%perform" | "%reperform"), _))) -> (
+          (not (List.is_empty r))
+          ||
+          match branch with
+          | Branch _ -> false
+          | Return x' -> not (Var.equal x x')
+          | _ -> true)
```

Comment:
Do we need to split for `Raise` and `Stop` ?

Reply:
Function `cps_block` only handles function calls and primitives followed by `Branch` and `Return`. For `Raise`, we have to generate code that is basically what we obtain by splitting anyway. It does not seem to be very useful to have a special case for `Stop`. I'm not even sure we can have a function call right before a `Stop`.